Outcome snippets docs

scholtzan opened the pull request:
This PR adds outcome snippet docs and also checks if I understand the specification correctly.

Basically, outcome snippets are like experiment configurations, except they do not have an `[experiment]` section and no analysis window specification for `metrics` (jetstream will compute them only for `weekly` and `overall` by default).

Closes https://github.com/mozilla/jetstream/issues/109.

scholtzan commented on README.md:
```diff
@@ -1,19 +1,23 @@
 Configuration files have four main sections:
-`[experiment]`, `[metrics]`, `[data_sources]`, and `[segments]`.
+[`[experiment]`](#experiment-section), [`[metrics]`](#metrics-section), [`[data_sources]`](#defining-data-sources),
+and [`[segments]`](#defining-segments).
```
I'm not sure about this one. Should outcome snippets allows to define segments? Or just metrics and data sources?

tdsmith replied:
Fascinating question. Let's say "no" for now; I can imagine some use cases but I don't think they're critical.

tdsmith commented:
I think we need some front-matter in the outcomes; we should provide a human-friendly name and description, at least.

tdsmith commented:
(Otherwise I think this seems right, thanks!)

scholtzan commented on README.md:
```diff
@@ -25,6 +29,21 @@ like `bug_12345_my_cool_experiment.toml`.
+are stored in the `outcomes/` directory and file names serve as unique identifiers.
+
+Configuration files have a set of [`[metrics]` definitions](#defining-metrics) and [`[data_sources]`](#defining-data-sources).
+A `friendly_name` and `description` are required at the top of the outcome snippet config file.
```
Do we need to add `applications` here to specify what applications an outcome snippet supports?

See https://github.com/mozilla/jetstream/pull/474#discussion_r569642670

```suggestion
A `friendly_name` and `description` are required at the top of the outcome snippet config file.

For specifying applications an outcome snippet computes metrics and statistics for, `application` is required to specify a list of supported `app_id`s. For example: `applications = ['firefox-desktop']` or `applications = ['org.mozilla.fenix', 'org.mozilla.fenix.nightly', 'org.mozilla.fennec.aurora']`
```

tdsmith replied:
I think we should be consistent about using app_id or app_name, but yes, I think we need to represent application somehow.

It might be useful to organize them in separate paths, e.g. fenix/my_outcome.toml or firefox_desktop/my_outcome.toml.

scholtzan replied:
Is it possible that some outcomes could target both Fenix and desktop? E.g. outcomes related to Gecko? (Or Fenix and iOS, who knows what the Nimbus future will look like ;)) So this might not be compatible with having outcome definitions in separate paths>

tdsmith replied:
I'm okay with duplicating the outcomes in that case. Right now, there's no chance of being able to reuse the definitions because the table structures are different. In some post-Glean world, that might look different, but the feature sets of the apps are different enough that I don't think we'll regret the duplication too deeply.

tdsmith reviewed (commented):
Can you add an example to the readme?

@jaredlockhart Can you comment on whether this satisfies https://github.com/mozilla/jetstream/issues/434 as a spec for the outcomes or would you like more detail somewhere?

A couple questions:
- Do filenames need to be unique per-platform? I think "no."
- Given that, do we need to describe how to canonically identify an outcome? How much of the filename ends up in the DTO?
- How do we remove a stale outcome from circulation? I think "delete the file" is natural but we should call out that consumers need to tolerate it.

scholtzan commented on README.md:
```diff
@@ -25,6 +29,22 @@ like `bug_12345_my_cool_experiment.toml`.
 
+### Outcome snippets
+
+Outcome snippets, which define a collection of summaries with a common theme (e.g. "performace", "Picture in Picture use"),
```
Moved to the end 👍

scholtzan commented:
I added some `example_config.toml` to `fenix/` and `firefox_desktop/` for Experimenter to have some files for testing. And I also added an example to the README

jaredlockhart reviewed (commented):
Great job @scholtzan this looks really good.  Thank you for adding the example files that's exactly what we need to start testing integration.

As @tdsmith pointed out I think there are a couple things worth considering: 

* If we use the filename as the outcomes slug, either we'll need
  * Each outcomes filename alone is globally unique (which ideally would be enforced here at CI validation time)
    * Related, even if all outcomes have a globally unique filename/slug, two outcomes might have the same `friendly_name` in which case there would be no way to distinguish them in the Experimenter UI 
  * The pair of `(application, slug)` is unique where application is one of [these values](https://github.com/mozilla/experimenter/blob/main/app/experimenter/experiments/constants/nimbus.py#L91-L93)
    * If the `application` labels don't match the constants in Experimenter we'll need a stable mapping
* Regardless of how we uniquely identify them, if outcomes here are applicable only to a single platform we'll need to do additional filtering on the Experimenter side to enforce that mapping (which is not yet specced)
* Whichever unique identifier an outcome has, it should be immutable once created, otherwise changing its identifier is the same as deleting that outcome and creating a new one and orphaning the records in Experimenter that point to the old identifier
* We'll have to deal with the case where an outcome slug in experimenter does not match any outcomes in the jetstream repo (possibly becuase the outcome was removed or renamed, or some other reason) and figure out what we want to display in Experimenter in that case

@tdsmith @scholtzan I think the critical thing to figure out before we land this is what an outcomes unique identifier is, and if it includes the application slug, ensuring those match Experimenters application slugs, then we should be able to move forward and address more of the outcome lifecycle concerns later.

I'd also like to get @emtwo 's feedback on this design.

emtwo commented:
This design looks good to me pending resolution of the naming convention for the slugs.

@jaredlockhart we hadn't included `description` in the breakdown we did on the experimenter side for outcomes work, only slug and name, but I think that would be a really great bonus to include.

> We'll have to deal with the case where an outcome slug in experimenter does not match any outcomes in the jetstream repo (possibly becuase the outcome was removed or renamed, or some other reason) and figure out what we want to display in Experimenter in that case

This sounds like a bit of an awkward case if someone chose an outcome for their experiment and then it gets removed while they're running it. For starters we could perhaps manually check if any experiments are currently live and referencing an outcome that needs to be removed before we remove it, but we could address this better later on.

> Related, even if all outcomes have a globally unique filename/slug, two outcomes might have the same friendly_name in which case there would be no way to distinguish them in the Experimenter UI

Perhaps at CI validation time the uniqueness of outcome name can be enforced too, at least per application.

tdsmith commented:
Outcomes are only applicable to a single platform.

> The pair of (application, slug) is unique where application is one of these values. If the application labels don't match the constants in Experimenter we'll need a stable mapping

We're using the `app_name` values that we expect in the DTO; can we standardize on those? I see the value `firefox-desktop` at that link; note that we're expecting the spelling `firefox_desktop`.

I think this is the right uniqueness constraint, fwiw.

> Whichever unique identifier an outcome has, it should be immutable once created, otherwise changing its identifier is the same as deleting that outcome and creating a new one and orphaning the records in Experimenter that point to the old identifier

Outcomes are mutable, which is why we're going to publish information about the version Jetstream sees in our analysis metadata. The visualization layer should rely on the analysis metadata to understand how outcomes relate to metrics, not the current contents of jetstream-config.

> We'll have to deal with the case where an outcome slug in experimenter does not match any outcomes in the jetstream repo (possibly becuase the outcome was removed or renamed, or some other reason) and figure out what we want to display in Experimenter in that case

Yeah, I think Experimenter should tolerate outcomes disappearing.

jaredlockhart reviewed (approved):
> Outcomes are only applicable to a single platform.
> 

Perfect 👍 

> > The pair of (application, slug) is unique where application is one of these values. If the application labels don't match the constants in Experimenter we'll need a stable mapping
> 
> We're using the `app_name` values that we expect in the DTO; can we standardize on those? I see the value `firefox-desktop` at that link; note that we're expecting the spelling `firefox_desktop`.

Ah yep we have that https://github.com/mozilla/experimenter/blob/main/app/experimenter/experiments/constants/nimbus.py#L96
> 
> I think this is the right uniqueness constraint, fwiw.
> 

Perfect 👍 

> > Whichever unique identifier an outcome has, it should be immutable once created, otherwise changing its identifier is the same as deleting that outcome and creating a new one and orphaning the records in Experimenter that point to the old identifier
> 
> Outcomes are mutable, which is why we're going to publish information about the version Jetstream sees in our analysis metadata. The visualization layer should rely on the analysis metadata to understand how outcomes relate to metrics, not the current contents of jetstream-config.
> 
> > We'll have to deal with the case where an outcome slug in experimenter does not match any outcomes in the jetstream repo (possibly becuase the outcome was removed or renamed, or some other reason) and figure out what we want to display in Experimenter in that case
> 
> Yeah, I think Experimenter should tolerate outcomes disappearing.

Okay after discussing this with @k88hudson @emtwo @tdsmith I think I have a clearer understanding of the flow:

1) At container build time, Experimenter snapshots the state of the outcome repo and uses that to pouplate the experiment creation form and stores the selected outcome slugs
2) At jetstream analysis time after the experiment ends, jetstream pulls in the experiment DTO with some set of outcome slugs, and the current set of outcomes from jetstream-config, and tries its best to match the outcome slugs in the DTO against the current set from the repo, some of which may appear in the DTO but not in the repo (in which cases it just silently passes over them) or the outcomes in the repo may have the same slug but have changed since they were selected at experiment creation time, in which case jetstream doesn't care and just uses the current version
3) When jetstream writes its analysis output and metadata to GCS it includes in the metadata the list of outcomes it analyzed, their slugs, names, descriptions, and the commit hash of the version of the outcome definition it used for its analysis, which should allow us to retrieve the exact state of the outcome as it was used during that analysis

I think this covers our bases for now.  We can address further complexity as it arises.  LGTM 👍
